azure/services/subnets/spec.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subnets
+
+import (
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+)
+
+// SubnetSpec defines the specification for a Subnet.
+type SubnetSpec struct {
+	Name              string
+	ResourceGroup     string
+	SubscriptionID    string
+	CIDRs             []string
+	VNetName          string
+	IsVNetManaged     bool
+	RouteTableName    string
+	SecurityGroupName string
+	Role              infrav1.SubnetRole
+	NatGatewayName    string
+}
+
+// ResourceName returns the name of the subnet.
+func (s *SubnetSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *SubnetSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName returns the name of the VNet that owns this subnet.
+func (s *SubnetSpec) OwnerResourceName() string {
+	return s.VNetName
+}
+
+// Parameters returns the parameters for the subnet.
+func (s *SubnetSpec) Parameters(existing interface{}) (parameters interface{}, err error) {
+	if existing != nil {
+		existingSubnet, ok := existing.(network.Subnet)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.Subnet", existing)
+		}
+
+		// If the subnet infrav1 spec has no CIDRs but the existing subnet has CIDRs, update the spec
+		if len(s.CIDRs) == 0 { // empty slice or nil?
+			// This is a bit of a workaround. By returning, we force a CreateOrUpdate call, which will trigger
+			// UpdateSubnetSpec() to be called. That will update the infrav1 spec, and those changes will be
+			// reflected in this spec when SubnetSpecs() is called.
```

Comment:
not sure I'm following what's going on here

when are we expecting this situation to arise?

Reply:
This is meant to replicate the behavior that was previously in the Reconcile loop. Before, if the subnet existed, we would get the resource and use it to update the ID and CIDRBlocks in the infrav1 subnet spec.

```
existingSubnet, err := s.getExisting(ctx, s.Scope.Vnet().ResourceGroup, subnetSpec)
switch {
case err != nil && !azure.ResourceNotFound(err):
    return errors.Wrapf(err, "failed to get subnet %s", subnetSpec.Name)
case err == nil:
    // subnet already exists, update the spec and skip creation
    s.Scope.SetSubnet(*existingSubnet)
    continue
```
```
// getExisting provides information about an existing subnet.
func (s *Service) getExisting(ctx context.Context, rgName string, spec azure.SubnetSpec) (*infrav1.SubnetSpec, error) {
	ctx, _, done := tele.StartSpanWithLogger(ctx, "subnets.Service.getExisting")
	defer done()

	subnet, err := s.Client.Get(ctx, rgName, spec.VNetName, spec.Name)
	if err != nil {
		return nil, errors.Wrapf(err, "failed to fetch subnet named %s in vnet %s", spec.VNetName, spec.Name)
	}

	var addresses []string
	if subnet.SubnetPropertiesFormat != nil && subnet.SubnetPropertiesFormat.AddressPrefix != nil {
		addresses = []string{to.String(subnet.SubnetPropertiesFormat.AddressPrefix)}
	} else if subnet.SubnetPropertiesFormat != nil && subnet.SubnetPropertiesFormat.AddressPrefixes != nil {
		addresses = to.StringSlice(subnet.SubnetPropertiesFormat.AddressPrefixes)
	}

	subnetSpec := s.Scope.Subnet(spec.Name)
	subnetSpec.ID = to.String(subnet.ID)
	subnetSpec.CIDRBlocks = addresses

	return &subnetSpec, nil
}
```


The old Reconcile logic doesn't explain why this is done, but I believe it's to auto fill the spec with the CIDR blocks for a custom vnet when the user only provides the subnet names, i.e. the example below. What do you think? 
```
networkSpec:
  subnets:
  - name: ${AZURE_CUSTOM_VNET_NAME}-controlplane-subnet
    role: control-plane
  - name: ${AZURE_CUSTOM_VNET_NAME}-node-subnet
    role: node
  vnet:
    name: ${AZURE_CUSTOM_VNET_NAME}
    resourceGroup: ${AZURE_RESOURCE_GROUP}
```

Reply:
> The old Reconcile logic doesn't explain why this is done, but I believe it's to auto fill the spec with the CIDR blocks for a custom vnet when the user only provides the subnet names, i.e. the example below. What do you think?
 
 that is correct
 
 I don't quite follow why we need to trick the reconciler and "force a CreateOrUpdate call" though. Wouldn't CreareResource return the existing resource if parameters returns nil already exactly for this use case?

Reply:
Oh that makes sense, in that case it looks like that's redundant.